csrc/device_lower/pass/alias_memory.cpp

```diff
@@ -1385,62 +1409,350 @@ class AllocationReuseModifier : private kir::ExprMutator {
+    alloc_stack_.push_back(alloc_info);
+  }
 
+  void assignNextAddress(AllocationInfo* alloc_info) {
+    auto alloc = alloc_info->alloc_expr;
+    if (alloc_stack_.empty()) {
+      alloc->setAddress(FusionGuard::getCurFusion()->zeroVal());
+    } else {
+      auto top_alloc = alloc_stack_.back()->alloc_expr;
+      auto top_size = allocSizeBytes(top_alloc);
+      auto unaligned_address =
+          SimplifyingIrBuilder::addExpr(top_alloc->address(), top_size);
+      auto aligned_address = alignExpr(unaligned_address);
+      // TODO: hoisting of addresses using for_loops_ recorded at first write
+      alloc->setAddress(aligned_address);
+    }
     if (isDebugDumpEnabled(DebugDumpOption::BufferReuseInfo)) {
-      debug() << "Allocated address " << address->toInlineString()
-              << " of size " << size_bytes->toInlineString() << " for T"
-              << alloc->buffer()->name() << " which has dtype "
-              << alloc->buffer()->dtype() << std::endl;
+      debug() << "Allocated address " << alloc->address()->toInlineString()
+              << " for T" << alloc->buffer()->name() << std::endl;
     }
+  }
 
-    current_address_ = SimplifyingIrBuilder::addExpr(address, size_bytes);
+  //! Record first reads and last writes, respecting aliased buffers
+  void recordEvents() {
+    for (auto& alloc_info : allocation_info_map_.allAllocationInfos()) {
+      if (alloc_info->alias_to) {
+        auto alias_info =
+            allocation_info_map_.getMaybeAllocationInfo(alloc_info->alias_to);
+        TORCH_CHECK(alias_info.has_value());
+        auto prev_last_read = lastAliasedRead(alias_info.value());
+        last_aliased_read_[alias_info.value()] = std::max(
+            prev_last_read, alloc_info->outer_live_interval->lastRead());
+      } else {
+        last_aliased_read_[alloc_info.get()] =
+            alloc_info->outer_live_interval->lastRead();
+      }
+    }
+
+    for (auto [alloc_info, last_read_pos] : last_aliased_read_) {
+      // Record the first write
+      auto write_it =
+          first_write_positions_
+              .emplace(alloc_info->outer_live_interval->firstWrite(), 0)
+              .first;
+      write_it->second.push_back(alloc_info);
+      // Ensure there is an entry for the last read position
+      last_read_positions_.insert(last_read_pos);
+    }
+  }
+
+  //! Pop all allocations on the top of the stack that are no longer active.
+  void reclaimMemory() {
+    if (!waiting_to_push_.empty()) {
+      // Check whether we have any allocations waiting to be pushed that are now
+      // inactive. If so, then they need to be ordered and allocated at the top
+      // of the stack before continuing.
+      if (std::any_of(
+              waiting_to_push_.begin(),
+              waiting_to_push_.end(),
+              [this](AllocationInfo* alloc_info) {
+                return lastAliasedRead(alloc_info) <= position_;
+              })) {
+        sortPushAndAssignWaiting();
```

Comment:
In the comment, it says sort and push at last read, but we are actually doing the sort and push right at sync.

Reply:
I'm not sure which comment you're referring to. This should run whenever we try to reclaim memory, which as you say is at every syncing expr.

Reply:
I am referring to this:
```C++
//! pushing/allocating whenever we encounter a first write, we can instead
//! append the allocation into a holding area vector. When we next encounter a
//! last read for some allocation, we can check whether it is in the holding
//! area. If so, then we form a mini-stack made of only the holding area
```
which says we are doing `sortPushAndAssignWaiting`  at each last read, instead of at sync point.

Reply:
Ah you're right. I need to rework that paragraph. Thanks for pointing that out.